File: without-library/graph/shortest_path/dijkstra.py
import sys

# Dijkstra with Manual Min-Heap

class MinHeap:
    def __init__(self):
        self.heap = []
        
    def push(self, item):
        self.heap.append(item)
        self._sift_up(len(self.heap) - 1)
        
    def pop(self):
        if not self.heap:
            return None
        min_item = self.heap[0]
        last_item = self.heap.pop()
        if self.heap:
            self.heap[0] = last_item
            self._sift_down(0)
        return min_item
        
    def _sift_up(self, idx):
        parent = (idx - 1) // 2
        if idx > 0 and self.heap[idx] < self.heap[parent]:
            self.heap[idx], self.heap[parent] = self.heap[parent], self.heap[idx]
            self._sift_up(parent)
            
    def _sift_down(self, idx):
        smallest = idx
        left = 2 * idx + 1
        right = 2 * idx + 2
        
        if left < len(self.heap) and self.heap[left] < self.heap[smallest]:
            smallest = left
        if right < len(self.heap) and self.heap[right] < self.heap[smallest]:
            smallest = right
            
        if smallest != idx:
            self.heap[idx], self.heap[smallest] = self.heap[smallest], self.heap[idx]
            self._sift_down(smallest)

    def is_empty(self):
        return len(self.heap) == 0

class Dijkstra:
    def __init__(self, n, adj):
        self.n = n
        self.adj = adj
        self.INF = 10**18
        self.dist = [self.INF] * (n + 1)
        
    def solve(self, start):
        self.dist[start] = 0
        pq = MinHeap()
        pq.push((0, start))
        
        while not pq.is_empty():
            d, u = pq.pop()
            
            if d > self.dist[u]:
                continue
                
            for v, w in self.adj[u]:
                if self.dist[u] + w < self.dist[v]:
                    self.dist[v] = self.dist[u] + w
                    pq.push((self.dist[v], v))

if __name__ == "__main__":
    pass

File: without-library/graph/shortest_path/dijkstra.cpp
#include <iostream>
#include <vector>

using namespace std;

typedef long long ll;
const ll INF = 1e18;

// Manual Min-Heap
struct MinHeap {
    vector<pair<ll, int>> heap;
    
    void push(pair<ll, int> val) {
        heap.push_back(val);
        sift_up(heap.size() - 1);
    }
    
    pair<ll, int> pop() {
        pair<ll, int> res = heap[0];
        heap[0] = heap.back();
        heap.pop_back();
        if (!heap.empty()) sift_down(0);
        return res;
    }
    
    bool empty() {
        return heap.empty();
    }
    
    void sift_up(int idx) {
        while (idx > 0) {
            int parent = (idx - 1) / 2;
            if (heap[idx] < heap[parent]) {
                pair<ll, int> temp = heap[idx];
                heap[idx] = heap[parent];
                heap[parent] = temp;
                idx = parent;
            } else {
                break;
            }
        }
    }
    
    void sift_down(int idx) {
        while (true) {
            int smallest = idx;
            int left = 2 * idx + 1;
            int right = 2 * idx + 2;
            
            if (left < heap.size() && heap[left] < heap[smallest])
                smallest = left;
            if (right < heap.size() && heap[right] < heap[smallest])
                smallest = right;
                
            if (smallest != idx) {
                pair<ll, int> temp = heap[idx];
                heap[idx] = heap[smallest];
                heap[smallest] = temp;
                idx = smallest;
            } else {
                break;
            }
        }
    }
};

struct Edge {
    int to;
    int weight;
};

vector<Edge> adj[200005];
ll dist[200005];

void dijkstra(int start, int n) {
    for(int i=0; i<=n; i++) dist[i] = INF;
    
    MinHeap pq;
    dist[start] = 0;
    pq.push({0, start});
    
    while(!pq.empty()) {
        pair<ll, int> top = pq.pop();
        ll d = top.first;
        int u = top.second;
        
        if (d > dist[u]) continue;
        
        for(size_t i=0; i<adj[u].size(); i++) {
            int v = adj[u][i].to;
            int w = adj[u][i].weight;
            
            if (dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w;
                pq.push({dist[v], v});
            }
        }
    }
}

int main() {
    return 0;
}
